hypothesis-python/src/hypothesis/core.py

```diff
@@ -768,6 +771,45 @@ def execute(data, function):
     return default_executor
 
 
+@contextlib.contextmanager
+def unwrap_exception_group():
+    def _flatten_group(excgroup: BaseExceptionGroup) -> list[BaseException]:
+        found_exceptions = []
+        for exc in excgroup.exceptions:
+            if isinstance(exc, BaseExceptionGroup):
+                found_exceptions.extend(_flatten_group(exc))
+            else:
+                found_exceptions.append(exc)
+        return found_exceptions
+
+    try:
+        yield
+    except BaseExceptionGroup as excgroup:
+        # Found? RewindRecursive? FlakyReplay?
+        marker_exceptions = (StopTest, UnsatisfiedAssumption)
+        frozen_exceptions, non_frozen_exceptions = excgroup.split(Frozen)
+        marker_exceptions, user_exceptions = non_frozen_exceptions.split(lambda e: isinstance(e, marker_exceptions))
+
+        # TODO: not a great variable name
+        if user_exceptions:
+            raise
+
+        if non_frozen_exceptions:
+            flattened_non_frozen_exceptions = _flatten_group(non_frozen_exceptions)
+            if len(flattened_non_frozen_exceptions) == 1:
+                raise flattened_non_frozen_exceptions[0] from None
+            # multiple non-frozen exceptions, re-raise the entire group
+            raise
+
+        flattened_frozen_exceptions = _flatten_group(frozen_exceptions)
+        # we only have frozen exceptions
+        if len(flattened_frozen_exceptions) == 1:
+            raise flattened_frozen_exceptions[0] from excgroup
+
+        # multiple frozen exceptions
+        # TODO: raise a group? The first one? None of them?
+        raise frozen_exceptions[0] from excgroup
+
```

Comment:
> If all exceptions are Frozen, raise the first (as above).

Actually, I think this could cause confusion. If I spawn three tasks and I only get one `Frozen`, I will assume that the-thing-that-caused-Frozen only happened in one of the tasks. We should probably err on the side of not unwrapping unless necessary to get internal exception handling to work.

Reply:
Another question is if we get multiple `HypothesisException`/`StopException`, if that's possible?

Reply:
If we get multiple `StopTest`, re-raise the one with the lowest _testcounter i.e. the first one to occur.